File: 1386-cinema-seat-allocation/1386-cinema-seat-allocation.cpp
class Solution {
public:
    int maxNumberOfFamilies(int n, vector<vector<int>>& reservedSeats) {
        unordered_map<int, int> reserved;

        for(const auto& seat : reservedSeats){
            int row = seat[0];
            int col = seat[1];

            if (col > 1 && col < 10){
                reserved[row] |= (1 << col);
            }
        }

        int maxFamilies = (n - reserved.size()) * 2;

        for(const auto& [row, bitmask] : reserved){
            bool leftFree = (bitmask & 60) == 0;
            bool rightFree = (bitmask & 960) == 0;
            bool middleFree = (bitmask & 240) == 0;

            if (leftFree && rightFree){
                maxFamilies += 2;
            } else if (leftFree || rightFree || middleFree){
                maxFamilies += 1;
            }
        }

        return maxFamilies;
    }
};

File: 1386-cinema-seat-allocation/1386-cinema-seat-allocation.py
import collections
from typing import List

class Solution:
    def maxNumberOfFamilies(self, n: int, reservedSeats: List[List[int]]) -> int:
        reserved = collections.defaultdict(int)

        for row, seat in reservedSeats:
            reserved[row] |= (1 << seat)

        max_families = (n - len(reserved)) * 2

        for bitmask in reserved.values():
            left_free = (bitmask & 60) == 0
            right_free = (bitmask & 960) == 0
            middle_free = (bitmask & 240) == 0

            if left_free and right_free:
                max_families += 2
            elif left_free or right_free or middle_free:
                max_families += 1

        return max_families